### ISSUE TITLE
eventmachine: Installation fails on two issues

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1541]
Hello,
I'm getting the following two issues when I'm tryting to compile eventmachine due to some Rack specs

```
$ gem install eventmachine
Building native extensions.  This could take a while...
ERROR:  Error installing eventmachine:
	ERROR: Failed to build gem native extension.

    current directory: ~/.gem/truffleruby/2.4.4/gems/eventmachine-1.2.7/ext
~/.rubies/truffleruby-1.0.0-rc11/bin/truffleruby -r ./siteconf20190121-55140-kw3wbq.rb extconf.rb
checking for -lcrypto... yes
checking for -lssl... yes
checking for openssl/ssl.h... no
checking for -lcrypto... yes
checking for -lssl... yes
checking for openssl/ssl.h... no
checking for -lcrypto... yes
checking for -lssl... yes
checking for openssl/ssl.h... no
checking for -lcrypto... yes
checking for -lssl... yes
checking for openssl/ssl.h... yes
checking for openssl/err.h... yes
-----
Found OpenSSL in path /usr/local/opt/openssl
-----
checking for rb_trap_immediate in ruby.h,rubysig.h... no
checking for rb_thread_blocking_region()... no
checking for rb_thread_call_without_gvl() in ruby/thread.h... yes
checking for rb_thread_fd_select()... yes
checking for rb_fdset_t in ruby/intern.h... yes
checking for rb_wait_for_single_fd()... yes
checking for rb_enable_interrupt()... no
checking for rb_time_new()... yes
checking for inotify_init() in sys/inotify.h... no
checking for __NR_inotify_init in sys/syscall.h... no
checking for writev() in sys/uio.h... yes
checking for pipe2() in unistd.h... no
checking for accept4() in sys/socket.h... no
checking for SOCK_CLOEXEC in sys/socket.h... no
checking for sys/event.h... yes
checking for sys/queue.h... yes
checking for clock_gettime()... yes
checking for CLOCK_MONOTONIC_RAW in time.h... yes
checking for CLOCK_MONOTONIC in time.h... yes
CXXFLAGS=-Xclang -disable-O0-optnone -g -Wimplicit-function-declaration -Wno-unknown-warning-option -Wno-int-conversion -Wno-int-to-pointer-cast -Wno-unused-value -Wno-incompatible-pointer-types -ferror-limit=500 -c -emit-llvm -stdlib=libc++ -Wall -Wextra -Wno-deprecated-declarations -Wno-ignored-qualifiers -Wno-unused-result -Wno-address
creating Makefile

current directory: ~/.gem/truffleruby/2.4.4/gems/eventmachine-1.2.7/ext
make "DESTDIR=" clean

current directory: ~/.gem/truffleruby/2.4.4/gems/eventmachine-1.2.7/ext
make "DESTDIR="
compiling binder.cpp
In file included from binder.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:810:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct RBasic {
^
In file included from binder.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1777:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:19: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                  ^~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:36: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                                   ^~~~~~~~~
In file included from binder.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:37:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/io.h:24:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/encoding.h:280:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct rb_econv_t {};
^
4 warnings generated.
compiling cmain.cpp
In file included from cmain.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:810:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct RBasic {
^
In file included from cmain.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1777:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:19: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                  ^~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:36: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                                   ^~~~~~~~~
In file included from cmain.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:37:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/io.h:24:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/encoding.h:280:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct rb_econv_t {};
^
4 warnings generated.
compiling ed.cpp
In file included from ed.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:810:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct RBasic {
^
In file included from ed.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1777:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:19: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                  ^~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:36: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                                   ^~~~~~~~~
In file included from ed.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:37:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/io.h:24:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/encoding.h:280:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct rb_econv_t {};
^
4 warnings generated.
compiling em.cpp
In file included from em.cpp:23:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:810:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct RBasic {
^
In file included from em.cpp:23:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1777:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:19: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                  ^~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:36: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                                   ^~~~~~~~~
In file included from em.cpp:23:
In file included from ./project.h:168:
In file included from ./em.h:37:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/io.h:24:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/encoding.h:280:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct rb_econv_t {};
^
4 warnings generated.
compiling kb.cpp
In file included from kb.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:810:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct RBasic {
^
In file included from kb.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1777:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:19: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                  ^~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:36: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                                   ^~~~~~~~~
In file included from kb.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:37:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/io.h:24:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/encoding.h:280:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct rb_econv_t {};
^
4 warnings generated.
compiling page.cpp
In file included from page.cpp:21:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:810:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct RBasic {
^
In file included from page.cpp:21:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1777:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:19: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                  ^~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:36: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                                   ^~~~~~~~~
In file included from page.cpp:21:
In file included from ./project.h:168:
In file included from ./em.h:37:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/io.h:24:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/encoding.h:280:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct rb_econv_t {};
^
4 warnings generated.
compiling pipe.cpp
In file included from pipe.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:810:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct RBasic {
^
In file included from pipe.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1777:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:19: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                  ^~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:36: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                                   ^~~~~~~~~
In file included from pipe.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:37:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/io.h:24:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/encoding.h:280:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct rb_econv_t {};
^
4 warnings generated.
compiling rubymain.cpp
In file included from rubymain.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:810:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct RBasic {
^
In file included from rubymain.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:24:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby.h:33:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1777:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:19: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                  ^~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/intern.h:47:36: warning: 'register' storage class specifier is deprecated and incompatible with C++17 [-Wdeprecated-register]
void rb_mem_clear(register VALUE*, register long);
                                   ^~~~~~~~~
In file included from rubymain.cpp:20:
In file included from ./project.h:168:
In file included from ./em.h:37:
In file included from ~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/io.h:24:
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/encoding.h:280:1: warning: empty struct has size 0 in C, size 1 in C++ [-Wextern-c-compat]
struct rb_econv_t {};
^
rubymain.cpp:1220:8: error: assigning to 'VALUE' (aka 'void *') from incompatible type 'int'
        arg = (NIL_P(arg)) ? -1 : NUM2INT (arg);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rubymain.cpp:1221:18: error: no matching function for call to 'evma_set_rlimit_nofile'
        return INT2NUM (evma_set_rlimit_nofile (arg));
                        ^~~~~~~~~~~~~~~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1464:31: note: expanded from macro 'INT2NUM'
#define INT2NUM(x) RB_INT2NUM(x)
                              ^
~/.rubies/truffleruby-1.0.0-rc11/lib/cext/include/ruby/ruby.h:1441:41: note: expanded from macro 'RB_INT2NUM'
# define RB_INT2NUM(v) RB_INT2FIX((int)(v))
                                        ^
./eventmachine.h:131:6: note: candidate function not viable: cannot convert argument of incomplete type 'VALUE' (aka 'void *') to 'int' for 1st argument
        int evma_set_rlimit_nofile (int n_files);
            ^
4 warnings and 2 errors generated.
make: *** [rubymain.bc] Error 1

make failed, exit code 2

Gem files will remain installed in ~/.gem/truffleruby/2.4.4/gems/eventmachine-1.2.7 for inspection.
Results logged to ~/.gem/truffleruby/2.4.4/extensions/x86_64-darwin/2.4.0/eventmachine-1.2.7/gem_make.out
```

**Comment by eregon** [comment]
The problematic code is here:
https://github.com/eventmachine/eventmachine/blob/8f7e1286be6aa05ec7e183e0cfa9381f6cf37e29/ext/rubymain.cpp#L1236-L1240
```c
static VALUE t_set_rlimit_nofile (VALUE self UNUSED, VALUE arg)
{
	arg = (NIL_P(arg)) ? -1 : NUM2INT (arg);
	return INT2NUM (evma_set_rlimit_nofile (arg));
}
```
`arg` is used both as a `VALUE` and an `int`.
This accidentally works on MRI because `VALUE` is typed as `long`.
I think in this case the source code should be modified to use a different variable for the `int` argument passed to `evma_set_rlimit_nofile`.

**Comment by eregon** [comment]
I opened a PR to `eventmachine` to address this: https://github.com/eventmachine/eventmachine/pull/878

**Comment by eregon** [comment]
PR merged upstream, now waiting for an `eventmachine` release.

**Comment by eregon** [comment]
With 0552ac72e3396492dc42ea53c1ac5ee2eb24a3ba `eventmachine` 1.2.7 almost installs, but fails on 
```c
VALUE rb_mProcess = rb_const_get(rb_cObject, rb_intern("Process"));
```
Because `rb_mProcess` is a macro for us (and not a global variable). I'll try to fix that.

**Comment by eregon** [comment]
This is now fixed with 0552ac72e3396492dc42ea53c1ac5ee2eb24a3ba and a181f79eefd53b6f2c4a6f060bcd361466bec0c0.
`gem install eventmachine` works,
and `ruby -e 'require "eventmachine"; p EM; EM.run { puts "OK" }'` seems to work too!
